relay-quotas/src/redis.rs

```diff
@@ -125,6 +139,291 @@ impl std::ops::Deref for RedisQuota<'_> {
+        client: &mut PooledClient,
+        quantity: usize,
+    ) -> Result<(usize, usize), GlobalRateLimitError> {
+        // We don't want to reject a metrics batch just because it's too large, so we increase the
+        // requested budget if the default is less than the metric bucket quantity.
+        let requested_budget = std::cmp::max(DEFAULT_BUDGET_REQUEST, quantity);
+        let redis_key = quota.key();
+        let expiry = quota.timestamp.as_secs() + quota.window + GRACE;
+
+        let (budget, redis_count): (usize, usize) = self
+            .script
+            .prepare_invoke()
+            .key(redis_key.as_str())
+            .arg(quota.limit)
+            .arg(expiry)
+            .arg(requested_budget)
+            .invoke(
+                &mut client
+                    .connection()
+                    .map_err(|_| GlobalRateLimitError::Redis)?,
+            )
+            .map_err(|_| GlobalRateLimitError::Redis)?;
+
+        Ok((budget, redis_count))
+    }
+
+    /// Ensures the local counter is up to date with the redis counter.
+    fn redis_sync(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<(), GlobalRateLimitError> {
+        let shared_counter_map = self.counters.read().unwrap_or_else(|e| e.into_inner());
+        let current_slot = current_slot(quota.window) as usize;
+        let budget_key = BudgetKeyRef::new(quota);
+
+        match shared_counter_map.get(&budget_key) {
+            Some(val) => {
+                let lock_available = val.try_read().is_ok();
+
+                if lock_available {
+                    let mut exclusive_val_lock = val.write().unwrap_or_else(|e| e.into_inner());
+                    let (new_budget, redis_value) = self.call_redis(quota, client, quantity)?;
+                    exclusive_val_lock.update(new_budget, redis_value, current_slot);
+                } else {
+                    drop(val.write().unwrap_or_else(|e| e.into_inner()));
+                }
+            }
+            None => {
+                drop(shared_counter_map); // Necessary to avoid deadlock.
+
+                // This evaluating to false implies that another thread is currently in the process
+                // of calling redis and inserting a new key.
+                let lock_available = self.counters.try_read().is_ok();
+
+                if lock_available {
+                    let mut exclusive_map =
+                        self.counters.write().unwrap_or_else(|e| e.into_inner());
+
+                    let (new_budget, redis_value) = self.call_redis(quota, client, quantity)?;
+
+                    exclusive_map.insert(
+                        budget_key.into_owned(),
+                        RwLock::new(BudgetState::new(new_budget, redis_value, current_slot)),
+                    );
+                } else {
+                    drop(self.counters.write().unwrap_or_else(|e| e.into_inner()));
```

Comment:
the idea was that if another thread is currently writing, it means they are calling redis, so it would be better to return when that thread is done with calling redis. if you call .write() it'd block until that thread is finished updating the redis results. 
  
now i have a atomicbool flag instead though